### PR TITLE
Added skip journal load on first run

### PR DESCRIPTION
**Pull request by jvymazal**
When starting rsyslog with imjournal for first time it outputs
an error and plugin does not run because no state file exists yet.
Now it skips the loading and creates state file on first persist.

### CODE REVIEW

**Comment by jvymazal** [comment]
@rgerhards we already had this behavior few versions ago but it was accidentally dropped in  #2065 , with this commit I have brought it back while also making it explicit so it stays with future code changes. Current behavior requires you to start rsyslog with imjournal twice before it starts working while it outputs error on first time.

File: plugins/imjournal/imjournal.c [hunk 1]
```diff
@@ -578,6 +578,17 @@ loadJournalState(void)
...
 
+	/* if state file not exists (on very first run), skip */
+	if (access(cs.stateFile, F_OK|R_OK) == -1 && errno == ENOENT) {
+		if (cs.bIgnorePrevious) {
+			/* Seek to the very end of the journal and ignore all older messages. */
+			skipOldMessages();
+		}
+		LogMsg(errno, RS_RET_FILE_NOT_FOUND, LOG_NOTICE, "imjournal: No statefile exists, "
+				"%s will be created (ignore if this is first run)", cs.stateFile);
+		goto finalize_it;
```
Comment (rgerhards) [inline plugins/imjournal/imjournal.c]: please use macro ```FINALIZE``` here (it expands to the same, but is defined to work with the rest of the macros for exception handling).

**Comment by rgerhards** [comment]
> we already had this behavior few versions ago but it was accidentally dropped in #2065 , with this commit I have brought it back while also making it explicit so it stays with future code changes. Current behavior requires you to start rsyslog with imjournal twice before it starts working while it outputs error on first time.

good catch! Could we add a testbench test to *ensure* that it isn't broken again in the future? Or, in other words: why did the existing tests not catch it?  (my fault maybe?)

**Comment by jvymazal** [comment]
If I am looking correctly current imjournal tests (there are only 2 of them) do not specify a statefile in config so have no way to catch this.

**Comment by rgerhards** [comment]
> If I am looking correctly current imjournal tests (there are only 2 of them) do not specify a statefile in config so have no way to catch this.

ah, ok, that makes sense. Could you add a test that does that? I guess you can simply copy one of the others, at least it would detect when it does not start. Maybe a bit more elaborate testing of the state persistence would of course be better ;-)

**Comment by jvymazal** [comment]
Also, when I am going trough the current tests it looks more like it is checking that TESTMSG is present in journal (https://github.com/rsyslog/rsyslog/blob/fd2200f3b55568d9e50e1655bd3196b56e8c2429/tests/imjournal-basic.sh#L26) then in rsyslog.out.log, if that is the case it would not detect even if the module would not be working.

**Comment by rgerhards** [comment]
> Also, when I am going trough the current tests it looks more like it is checking that TESTMSG is present in journal :

yeah, that's right, but is it probably that this message (note the date part)
```
TESTMSG="TestBenCH-RSYSLog imjournal This is a test message - $(date +%s)"
```
is present without the test running?

Anyhow, suggestions on how to make it more bullet proof is highly recommended.

**Comment by rgerhards** [comment]
Doh! The test should go against rsyslog.out.log!

**Comment by jvymazal** [comment]
I have changed all the (now 3) imjournal tests to go against the right file as well as adding one that uses statefile so we have that covered. Unless you have anything else, merge whenever you are ready.

**Comment by rgerhards** [comment]
Oh, I see there were problems with the journal test. Makefile.am quoted https://github.com/rsyslog/rsyslog/issues/2564 - we probably need to check what was going on here. Also looks like the issue of test validity already came up there.

**Comment by jvymazal** [comment]
@rgerhards , I think I solved the problem for now, it is not the most elegant solution however it should give us stable state and it is still much better than having all imjournal tests disabled. From what I was able to find out in some cases journal input just does not work (possibly some permissions error), in imjournal tests I now detect those cases and skip the test. Also if you could please merge it when the rest of CI goes green so I can rebase #2543 onto it and continue working on it with working testbench. Thanks.

**Comment by rgerhards** [comment]
@jvymazal  sry for breaking your commit - I should have reversed merge order but have to admit I did not watch closely enough... Even more thx for your fast action!

**Comment by jvymazal** [comment]
@rgerhards now when finally it all works with the new testbench scripts I have gone trough the results and it looks like there are problems with journald on some of the docker environments (although it appears working journal db itself is empty), others are functional and tests are green now, so if you merge we can close #2564 by this.

**Comment by rgerhards** [comment]
That journal thing makes sense, as there is no systemd et all running in
the containers. Let me work on the image options tomorrow, I think we need
to disable testing on the containers (but can build).

Rainer

Sent from phone, thus brief.


jvymazal <notifications@github.com> schrieb am Di., 24. Juli 2018, 21:35:

> @rgerhards <https://github.com/rgerhards> now when finally it all works
> with the new testbench scripts I have gone trough the results and it looks
> like there are problems with journald on some of the docker environments
> (although it appears working journal db itself is empty), others are
> functional and tests are green now, so if you merge we can close #2564
> <https://github.com/rsyslog/rsyslog/issues/2564> by this.
>

**Comment by davidelang** [comment]
you have to do some extra trickery to get journald working in a container. I've 
set it up, but I had to google for it.

**Comment by rgerhards** [comment]
actually I am not sure I want it that way. It would just be for the journal
tests, and remove much of the simplicity IMHO. we'll probably spawn a VM
for these tests... Or do you think the other way around would be much
better?

Rainer

2018-07-24 21:58 GMT+02:00 David Lang <notifications@github.com>:

**Comment by davidelang** [comment]
a vm will be more reliable as it won't be interacting with the host system.

David Lang

**Comment by jvymazal** [comment]
@rgerhards @davidelang In short, should I remove the journal-check blocks from tests or leave them there?

**Comment by jvymazal** [comment]
> Well... assuming the journal is not present - do we then have `journalctl`? 

That is exactly the problem, we have, even the `systemd-cat` actually exits clean (with 0) so I had to do this explicit check (insert message, check it is there, skip (77) if not), and only do the rsyslog.out check if it is present in journal.

**Comment by rgerhards** [comment]
> That is exactly the problem, we have, even the systemd-cat actually exits clean (with 0) so I had to do this explicit check (insert message, check it is there, skip (77) if not), and only do the rsyslog.out check if it is present in journal.

ah, ok, I hadn't looked at the code yet. I think this is a valid solution. I think if the journalctl check fails, there is no way the test could work. Right? If so, I think the cure is actually quite valid. We just need to make sure we have a (dedicated) builder inside CI that ensure the tests are run.

Are we in agreement?

**Comment by jvymazal** [comment]
> We just need to make sure we have a (dedicated) builder inside CI that ensure the tests are run.

Agree, if all of the non-container runs are enough we should be good (I have gone trough them and it does not look like any of them are skipping the tests).

**Comment by jvymazal** [comment]
@rgerhards If we agreed that current solution is acceptable could you please merge it? I would like to continue work on #2543 for which it would be nice if I could rebase it onto this so I have functional CI and both changes in one place for testing. Many thanks

**Comment by rgerhards** [comment]
thx for the ping - wrangling with CI, so do not check that closely. Pls ping on any other urgent thing as well if I don't act fast enough!

**Pull request merged**